File: java/ankit_kanyal/devnest_problems/ReOrderList.java
// link- https://devsnest.in/algo-challenges/reorder-list?tab=question

static Node getMiddle(Node head){

Node slow = head;

Node fast = head;
    
while(fast != null && fast.next != null){
    slow = slow.next;
    fast = fast.next.next;
}

return slow;
}

static Node reverseLinkedList(Node head){

Node prev = null;

Node curr = head;

while(curr != null){
    Node next_node = curr.next;
    curr.next = prev;
    prev = curr;
    curr = next_node;
}       

return prev;
}

static Node solve(Node head){
//CODE HERE 

if (head == null || head.next == null){
     return head;
}
Node middle = getMiddle(head);
    
Node first = head;
Node second = reverseLinkedList(middle);
    
while(second.next != null){
    Node temp = first.next;
    first.next = second;
    first =  second;
    second = temp;
}

return head;


}

File: python/vedant_bhatnagar/DevsnestProblems_THA/Phase 1/reorderlist.py
def cutinhalf(head):
    if not head or not head.next:
        return None
    slow, fast = head, head.next
    while fast.next and fast.next.next:
        slow = slow.next
        fast = fast.next.next
    if fast.next:
        slow = slow.next
    t = slow.next
    slow.next = None
    return t
    
def reverse(head):
    if not head or not head.next:
        return head
    p,c,n = head, head.next, head.next.next
    while n:
        c.next = p
        p, c, n = c ,n , n.next
    head.next = None
    c.next = p
    return c

def solve(head):
    # CODE HERE
    h = cutinhalf(head)
    h = reverse(h)
    p = head

    while h:
        x = h.next
        h.next = p.next
        p.next = h
        p = p.next.next
        h = x
    return head